tests/gui/TestGuiPixmaps.cpp

```diff
@@ -42,19 +43,21 @@ void TestGuiPixmaps::testEntryIcons()
 
     // Test setting standard icon
     entry->setIcon(10);
-    auto pixmap = entry->iconPixmap();
+    auto pixmap = Icons::entryIconPixmap(entry);
     QCOMPARE(pixmap.cacheKey(), databaseIcons()->icon(10).cacheKey());
 
     // Test setting custom icon
     QUuid iconUuid = QUuid::createUuid();
     QImage icon(2, 1, QImage::Format_RGB32);
     icon.setPixel(0, 0, qRgb(0, 0, 0));
     icon.setPixel(1, 0, qRgb(0, 0, 50));
-    db->metadata()->addCustomIcon(iconUuid, icon);
+    db->metadata()->addCustomIcon(iconUuid, Icons::saveToBytes(icon));
 
     entry->setIcon(iconUuid);
-    pixmap = entry->iconPixmap();
-    QCOMPARE(pixmap.cacheKey(), db->metadata()->customIconPixmap(iconUuid).cacheKey());
+    pixmap = Icons::entryIconPixmap(entry);
+    // FIXME Looks like the cacheKey is not the same for 2 QPixmap instances, even if they have the
+    // same content
```

Comment:
@droidmonkey that's the only blocker left. It looks like `QPixmap` is generating a different cache key even if the content of the pixmap is the same. I could replace that check with a check of the content of the pixmaps, but that means the global  `QPixmapCache` might no longer be working in the case of custom icons. What do you think?

Reply:
So I made some more tests and it does appear that QT will generate a new cache key if the address of the content is changed, which means that 2 `QPixmap`s will have a different cache key even if they hold equivalent content. What we could do is stop relying on `cacheKey()` to populate the `QPixmapCache`, and instead generate a hash of the content for the cache key, but I'm not sure what exactly the performance consequences of that will be. Is it really more efficient to hash an image than to render it again? Another option would be to have a cache overlay over `Database`s instances, to make sure `QPixmap`s are only generated once for the same content, but that would require significantly more effort, and I don't find it super elegant.

Reply:
@phoerious  would like to get your opinion on that as well.

Reply:
This is just a test. How often would you expect identical content in real life? For the test do whatever dirty hack you can come up with, but for production code I think we can rely on cacheKey.

Reply:
@phoerious I think the goal of the test is to make sure that `cacheKey` will always return the same value so that the `DatabaseIcons` will cache the images correctly. See https://github.com/keepassxreboot/keepassxc/blob/develop/src/core/DatabaseIcons.cpp#L79 for details.

Reply:
Perhaps a somewhat useless check.

Reply:
I don't even understand the caching in `applyBadge`. We call `cacheKey` before applying the badges. I feel like it might overwrite an image in the cache with the equivalent image but with badges applied.

Reply:
Maybe I'm just being lazy but I'd feel like just removing the caching with `QPixmapCache`. In the worst case we could add caching again with the new structure if we notice performance problems in the future, but I kinda doubt it for small icons like this. But what do I know :shrug: